### DISCUSSION TITLE
Difficulty Adjust should snap sliders to modified gameplay values.

### DISCUSSION THREAD

**Discussion by JamieWright99** [ppy/osu#24641] [Ideas]
I think it would be good if there was a setting/toggle that makes the difficulty adjust mod have its sliders to be snapped to what I'm going to see in game rather than mods not visually effecting them. Currently if I have an AR 7 map and put DT on, the slider will stay at AR 7 despite becoming AR 9~ in game, and of course it's the same with the other values. It becomes a bit awkward as some values become decimal but I still think it's worth adding.

Here's a mock-up of what I mean, as I change the values in the DT mod the sliders will move to the in game values. (Edited)

![Difficulty Adjust2](https://github.com/ppy/osu/assets/102170503/4669ba23-a9e6-4d7c-85d2-6a8abd4eed6c)

I tried searching and this is potentially a duplicate of #18189 , however I felt this is more regarding specifically the slider behaviour inside difficulty adjust rather than just displaying information. Feel free to mark as answered if you feel it relates too much though!

**Comment by Rekunan** [comment]
I see big problem where the extended limits of DA only goes up to 11. What if someone wants to reach values beyond 11? What should be done then?

**Reply by JamieWright99** [reply]
I'm not saying to extend any limits, I'm saying that values on the slider should go to the in-game value. An AR 10 map would still be capped to AR 11 with a speed up.

**Reply by Rekunan** [reply]
That brings a UX issue then, if the sliders are linked together, does that mean changing DA changes the values after the rate or before? If it's after, then once again, DA's extended limits caps out at 11, not allowing any further. This means the normal cap, AR12 if you do AR11 on 2.0x, would become unachievable. If it's before, then it becomes a confusing mess of order of priority.

**Reply by JamieWright99** [reply]
I think you are misunderstanding. The sliders are not linked together, they are linked to the speed change. Currently on Stable and in Lazer an AR8 DT song becomes AR 9.67, all I am saying is that value should be shown in the slider. It would also be the same with AR11 DT, there is no way to go beyond that value so the slider would stay at AR 11 in that scenario.

I think we are talking past each other so hopefully others understand what I mean, I'll stop replying to this chain.

**Reply by Rekunan** [reply]
> I think we are talking past each other

I think so too, so let me restart by saying and asking this:
If someone turns both DT and DA on, the DA sliders show the values after the rate change. Like you said, if the map is AR8, the slider would show AR9.67 (if not exact then rounded). What happens then when someone moves the AR slider, let's say they move it to AR10 for example, is that the AR after DT, or before?

**Reply by monochrome22** [reply]
It would be after. The sliders would have to be limited to the highest and lowest value that can be set with the current speed settings. For example, with 0.50x Half Time, the AR slider would be from -10 to 9 and with 2.00x Double Time, it would be from  7 to 12.
Adjusting Difficulty Adjust sliders won't adjust the speed slider, but adjusting the speed slider would adjust the DA sliders. Except if the user has manually changed the DA sliders, then it will stay at that value, unless if the speed setting is changed to a point where the set value is not possible, then it will automatically adjust to the highest or lowest possible value with the new speed setting as the limit of the slider changes (and not automatically adjusted back if the speed setting is adjusted back). And if the adjusted value goes over 10, then "Extended Limits" would automatically toggle (and not untoggle automatically).
Would be good to have more precise values for the DA sliders then, so players can set AR 10.33 etc. Players would naturally want to play with the same values that they are used to or what another score had. Maybe allow 0.01 precision values when holding shift.
If confusion is an issue, then this setting could be a toggle like OP suggested, and the toggle only appears in DA if DT or HT is enabled. "Show adjusted values" maybe. Though with https://github.com/ppy/osu/pull/24642, it might actually be better to make this the default behavior and put the current behavior under a toggle, like "Show base values". The current behavior should still be available, because players coming from stable might not know what exact value they are supposed to set. Meaning, if someone wants to play AR9.5 DT, they might not know that the real AR of that is 10.67.

By limiting sliders to highest and lowest value that can be set, there shouldn't be an issue with clock time and whatnot, as said in https://github.com/ppy/osu/discussions/14170#discussioncomment-1152266. 
By limiting sliders, you wouldn't be forcing the Difficulty Adjust AR over speed adjusts, instead you would be automatically adjusting the AR behind the scenes so the final result would be what is displayed. Meaning, if someone sets AR 10 DA with 1.5x DT, behind the scenes it's actually AR 8.5.

Alternatively, you could limit AR to 11 (by automatically lowering AR behind the scenes to ensure it stays at 11) and allow Difficulty Adjust to use even higher approach times (lower AR) behind the scenes, this way you wouldn't need to limit slider values. Meaning, at 2.00x speed, the user would see AR0, but behind the scenes it's AR -10 or something. 
This would mean you would be able to use higher and lower ARs/ODs than what are available with "Show base values", but that should be fine, since showing base values during speed adjust would be somewhat of a legacy feature for the people who are used to that. The biggest issue is that the object density would change quite a bit, since there would be a big difference between AR0 nomod and AR0 2.00x DT.

**Reply by Rekunan** [reply]
> with 0.50x Half Time, the AR slider would be from -10 to 9

It'd actually be -15 to 9

> with 2.00x Double Time, it would be from 7 to 12

Would that mean that DA would automatically open up to allow AR12 and such? Cause that's my main worry from the start.

> Maybe allow 0.01 precision values when holding shift.

It'd be better to just have it automatically snap to the values instead of introducing an unnecessary precision level (ex: 1.5x would snap to ar10.2, ar10.27, ar10.33, etc).

Assuming OP also agrees with everything else, I don't have any more worries about this idea.

**Reply by monochrome22** [reply]
> It'd actually be -15 to 9

I don't know, I took values using https://github.com/ppy/osu/pull/24642.

> Would that mean that DA would automatically open up to allow AR12 and such?

Like I said, what you can set with DA would be determined by what speed settings you have, the available DA values would adjust in real time as you adjust the speed setting slider. AR12 would only be available if you set 2.00x speed. 

> It'd be better to just have it automatically snap to the values instead of introducing an unnecessary precision level (ex: 1.5x would snap to ar10.2, ar10.27, ar10.33, etc).

That wouldn't work if you set an AR in DA, then adjust the speed slider. Meaning, if you set AR 10.33 with 1.5x speed and then change it to 1.6x speed, 10.33 wouldn't be available.

**Reply by JamieWright99** [reply]
Just wanted to comment that I didn't know AR12 was a thing at all, I assumed even 2x speed DTHR AR11 would still be capped at AR11 in game, but apparently its faster. I'm sorry for causing this confusion, I got 100 days in this game but I was just unaware of this 😅 Sorry Rekunan for that

**Reply by Rekunan** [reply]
> I took values using https://github.com/ppy/osu/pull/24642.

Oh interesting, I was using https://mbmasher.github.io/osu-tools, so maybe it is -10 to 9.

> That wouldn't work if you set an AR in DA, then adjust the speed slider.

Shit, yeah, it wouldn't be exact anymore. Maybe in that case, it should try snapping to the next closest value? Def something to look over more, because I still dislike providing unnecessary precision if it can be avoided.

> I didn't know AR12 was a thing at all ... I'm sorry for causing this confusion

Apology accepted. I had a hunch that you didn't know, but wasn't too sure about it. Glad to know it was right, and that I taught something new.

**Reply by monochrome22** [reply]
> Maybe in that case, it should try snapping to the next closest value?

I wanted to avoid that, because having the AR constantly change a little bit from the value that the user set might be confusing and it feels like broken UX. 
Would need more opinions on this. 

By the way, another solution would be to simply display real AR and OD in mod select. So when trying to get a precise AR in Difficulty Adjust with speed changes applied, you would be looking at the real values instead of the number on the slider. I think the main issue people have with DA+DT is that they have no idea what the AR is going to be in game. Most people would probably be happy if they could just see the real values in mod select. 
The downside of this is that you would constantly need to adjust DA sliders when making speed changes to maintain the AR/OD you want. Unless if there's some toggle to maintain closest values when making speed changes.
The OP did mention they want the slider behavior changed instead of just displaying information, but as can be seen from my previous posts, changing the slider behavior is more complicated than just displaying information, which would also largely solve the problem. Up to developers if they think it's worth it to change slider behavior.

**Reply by Rekunan** [reply]
> feels like broken UX

It def does sound jank yeah.

> another solution would be to simply display real AR and OD in mod select

Honestly, that might be the better option, so we don't have to deal with 1.5x having 10.33 and 1.6x having 10.31 changing the slider constantly.

**Comment by peppy** [comment]
Why is DT changing circle size in the mock up?

**Reply by JamieWright99** [reply]
I just moved all the sliders along when making that gif, I didn't think about circle size not being a changing value lol. Of course only the appropriate attributes would move in a real implementation
